pkg/service/signal.go

```diff
@@ -0,0 +1,144 @@
+package service
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/livekit/livekit-server/pkg/routing"
+	"github.com/livekit/livekit-server/pkg/telemetry/prometheus"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/rpc"
+	"github.com/livekit/psrpc"
+)
+
+type SessionHandler func(
+	ctx context.Context,
+	roomName livekit.RoomName,
+	pi routing.ParticipantInit,
+	connectionID livekit.ConnectionID,
+	requestSource routing.MessageSource,
+	responseSink routing.MessageSink,
+) error
+
+type SignalServer struct {
+	server rpc.TypedSignalServer
+}
+
+func NewSignalServer(
+	nodeID livekit.NodeID,
+	region string,
+	bus psrpc.MessageBus,
+	sessionHandler SessionHandler,
+) (*SignalServer, error) {
+	s, err := rpc.NewTypedSignalServer(nodeID, &signalService{region, sessionHandler}, bus)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debugw("starting relay signal server", "topic", nodeID)
+	if err := s.RegisterRelaySignalTopic(nodeID); err != nil {
+		return nil, err
+	}
+
+	return &SignalServer{s}, nil
+}
+
+func NewDefaultSignalServer(
+	currentNode routing.LocalNode,
+	bus psrpc.MessageBus,
+	router routing.Router,
+	roomManager *RoomManager,
+) (r *SignalServer, err error) {
+	sessionHandler := func(
+		ctx context.Context,
+		roomName livekit.RoomName,
+		pi routing.ParticipantInit,
+		connectionID livekit.ConnectionID,
+		requestSource routing.MessageSource,
+		responseSink routing.MessageSink,
+	) error {
+		prometheus.IncrementParticipantRtcInit(1)
+		return roomManager.StartSession(ctx, roomName, pi, requestSource, responseSink)
+	}
+
+	return NewSignalServer(livekit.NodeID(currentNode.Id), currentNode.Region, bus, sessionHandler)
+}
+
+func (r *SignalServer) Stop() {
+	r.server.Kill()
+}
+
+type signalService struct {
+	region         string
+	sessionHandler SessionHandler
+}
+
+func (r *signalService) RelaySignal(stream psrpc.ServerStream[*rpc.RelaySignalResponse, *rpc.RelaySignalRequest]) (err error) {
+	ctx, cancel := context.WithCancel(psrpc.NewContextWithIncomingHeader(context.Background(), psrpc.IncomingHeader(stream.Context())))
```

Comment:
what would be in the header / context? a bit of comments 🙏

Reply:
this is for smuggling the auth info to the server for cloud - do we want to call ths out in oss?

Reply:
I think would be helpful to note that it's to pass along context such as auth info to the other instance.